packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -1416,7 +1419,12 @@ class OverlayPortalController {
   void show() {
     final _OverlayPortalState? state = _attachTarget;
     if (state != null) {
+      final bool wasShowing = isShowing;
```

Comment:
What's the expected behavior when a controller gets attached to a different OverlayPortal that has a different `isShowing` state?

Reply:
Also it seems listeners are only notified when the controller is attached to a State, what's the reason behind that?

Reply:
Hmmm good points.

For 1, I agree we should notify if reattaching changes the underlying state. I will include this in my next commit.

For 2, I think you could make good arguments for a couple of options.
I was modelling the behavior here more or less off other controllers like scroll controller where their state is not valid when they're not attached-it didn't occur to me to handle this case.

Here are approaches that come to mind to me for 2:
A. Make show and hide throw when the controller is not attached
B. Allow show and hide when unattached, notify on change

My personal preference is for A as I dislike code that fails silently and it mirrors other controller methods like `ScrollController.animateTo`, but I'm happy to go with either or any third option you prefer.

Reply:
> Make show and hide throw when the controller is not attached

IIRC a lot of tests in flutter/flutter do that. I think `isShowing` actually updates even when detached.

Reply:
So I started working on making ALL interaction (show, hide, isShowing) with the controller throw if the controller is not attached.

I think this is the best behavior and most in keeping with other Flutter widgets, but it looks like it breaks just about all of the tests and some controller logic becomes obsolete and should be removed. Before I go about this rewrite, I wanted to double check (wasn't sure from your last post):
are you all on board with this change? and/or should I just do a barebones change now and not fire the listener when it's unattached, and then in a follow up PR migrate to throw-if-unattached?

Reply:
I think calling `show`/`hide` when the controller is detached should be legal. For example, if the OverlyPortal is in a lazy list view then there's usually no guarantee that the OverlayPortal widget will be built.

Reply:
Since the controller already has an `isShowing` getter, does it make sense to implement the value notifier interface instead?

Reply:
Cool I've made it so the controller notifies even if it's not attached and I've added a test to verify this behavior.

Also I got Git Confused and the commit history is now a nightmare XD please ignore.

As for implementing the value notifier interface, I certainly could, but I'd be worried that we could easily down the line want to expand the controller to notify for more than just a single value and would want to stop implementing the value notifier interface. eg if we decided to publicly expose z index. So might as well not implement value notifier int he first place.
But I don't feel strongly either way-if you still think we should let me know and I'll add it.

Reply:
Ah makes sense to me. Alternatively instead of making OverlayPortal itself listenable we could add a `ValueNotifier` / turn `isShowing` to a `ValueNotifier`, up to you.

I'm a bit surprised that the leak tracker isn't complaining about undisposed controllers now it becomes listenable.

Reply:
Sorry getting this PR in has dragged out so much-I've got a lot of other things fighting for my attention.

For the git commit history, I couldn't find a clean way to rebase as now master is so far ahead of my feature branch. Unless you know a good git trick to save this I'll just create a new branch and PR.

As for handling when the controller is reattached to something with a different value, sorry I forgot that part! But before I implement it, I think there's a bit of a confusing contradiction in our API as we're planning it:
1. We want to be able to call `show` and `hide` while the controller is not attached
2. We want to notify if the controller is attached to a one portal with one value, then attached to a different portal with a different showing value

These two feel contradictory. If I call `show` on a controller, then attach it to a portal where `isShow` is currently false, I'd expect the controller value to override the portal value. Otherwise, calls to `show` while attached are just silently failing which feels confusing.
How about instead of notify if attach then reattach changes the `isShowing` value, we instead make the controller the source of truth, so controller now holds `isShowing` and an `OverlayPortal` doesn't have its own concept of `isShowing`-it reflects it's attached controller's value. All this said, I'm not sure what `ScollController` etc's behavior is in this context-do you already know what those widgets do when dettaching and attaching to a new widget? I

Will spin up a new PR with a clean commit history once you let me know which direction you think we should take!

Reply:
Sorry I missed the github notification. You could use a new local branch and set the remote branch to the one associated with this PR and I think that updates the pull request. Or delete the local branch and recreate under the same branch name.

> we instead make the controller the source of truth,

Yeah that sounds reasonable. 

I don't think OverlayPortalController currently have a `dispose` method. If you make the class listenable you may have to add one and the owner is now supposed to call `dispose`. That could be a breaking change.